### PR TITLE
fix(cypher): evaluate label tests and function calls in a WITH-attached WHERE

### LINKED ISSUE #299: WHERE attached to WITH is not evaluated: label tests admit every row, function calls exclude every row
## Summary

A `WHERE` attached to a `WITH` is not evaluated as a filter for two kinds of predicate, and it fails in **opposite directions** depending on which kind — one silently returns everything, the other silently returns nothing. Neither errors.

Reference implementation for every number below: **Neo4j 5.x community**, same fixture, same statements.

## Reproduction

```cypher
CREATE (f:Function {uid:'fn1'}), (a:CloudAction {action:'s3:GetObject'}), (w:Workload {id:'w1', name:'checkout'});
```

| Query | Neo4j 5 | NornicDB |
| --- | ---: | ---: |
| `MATCH (n) WITH n WHERE n:Workload RETURN count(*)` | 1 | **3** |
| `MATCH (n) WHERE n:Workload WITH n RETURN count(*)` | 1 | 1 |
| `MATCH (a:Workload) WITH a WHERE toUpper(a.name) = 'CHECKOUT' RETURN count(*)` | 1 | **0** |
| `MATCH (a:Workload) WITH a WHERE a.name = 'checkout' RETURN count(*)` | 1 | 1 |
| `MATCH (a:Workload) WITH a WHERE coalesce(a.name,'X') IS NULL RETURN count(*)` | 0 | **1** |

Rows 2 and 4 are controls: moving the predicate before the `WITH`, or comparing a bare property, both work. So this is specific to clause position combined with predicate kind.

It also affects path traversal, which matters because that is the shape a scoped whitelist uses:

```cypher
MATCH (s:Repository {id:'r1'})-[*1..3]->(impacted)
WITH impacted WHERE impacted:Workload OR impacted:CloudResource
RETURN impacted.id
```

returns every reachable node, not just the whitelisted labels.

## Cause

Two evaluators, one per execution path, share the same shape:

- `evaluateWithWhereCondition` (`pkg/cypher/match_rows.go:173`) — simple-MATCH path
- `evaluateWhereOnComputedRow` (`pkg/cypher/match_with_rel.go:712`) — path-traversal path

Each handles `IS NULL`/`IS NOT NULL` and six comparison operators, then ends with a pass-through returning `true`.

**Label tests** carry no comparison operator, so they reach that pass-through and admit every row — the filter does nothing.

**Function calls** are split on the operator, and the operand resolver separates variable from property by locating a dot. In `toUpper(a.name)` it finds the dot *inside the call* and looks up a variable named `toUpper(a`, which never resolves — so the comparison is false for every row and the filter excludes everything. `coalesce(...) IS NULL` fails equivalently through the `IS NULL` branch.

## Why it is worth prioritising

The failure is silent in both directions. A whitelist that admits everything produces over-broad results that look plausible; a predicate that excludes everything produces an empty result that looks like "no matches." Neither surfaces an error, so a correct-looking query returns wrong rows and nothing detects it.

We hit this in production on a governed read path where the ignored predicate was an authorization-adjacent label whitelist, and separately on a projection query that returned zero rows for months without any signal.

## Fix

PR #300 fixes both evaluators, with regression tests stating the Neo4j-measured expectation for each case. After it, all five rows above agree with Neo4j.

## Note for maintainers, not part of the fix

The pass-through for wholly unrecognised predicates is pinned as intentional by `TestEvaluateWithWhereCondition_DirectBranches` ("Unknown pattern defaults to pass-through"), so the PR leaves it alone. It is worth knowing that it makes an unsupported filter behave as *no filter* rather than as an error, which is how both defects above stayed invisible. Whether that default should change is your call, not something a bug-fix PR should decide.

Measured on `main` at `8abc2269` (reports v1.2.2) and on the `eshu-nornicdb-pr290:3722b483c02c` build.

### PR DESCRIPTION
**Pull request by linuxdynasty**
Fixes #299.

A `WHERE` attached to a `WITH` was not evaluated as a filter for two kinds of predicate, and it failed in **opposite directions** depending on which kind. Neither errored.

## Measured against Neo4j 5.x community

Same fixture, same statements, Neo4j as the reference implementation:

| Query | Neo4j 5 | Before | After |
| --- | ---: | ---: | ---: |
| `MATCH (n) WITH n WHERE n:Workload RETURN count(*)` | 1 | **3** | 1 |
| `MATCH (n) WHERE n:Workload WITH n RETURN count(*)` | 1 | 1 | 1 |
| `WITH a WHERE toUpper(a.name) = 'CHECKOUT'` | 1 | **0** | 1 |
| `WITH a WHERE a.name = 'checkout'` | 1 | 1 | 1 |
| `WITH a WHERE coalesce(a.name,'X') IS NULL` | 0 | **1** | 0 |

Rows 2 and 4 are controls and were already correct — the predicate moved before the `WITH`, and a bare property comparison. They are in the test file to keep the fix honest about what it changed.

## What was wrong

Two evaluators, one per execution path, share the same shape: handle `IS NULL`/`IS NOT NULL` and six comparison operators, then fall through to `return true`.

**Label tests** carry no comparison operator, so they reached that pass-through and admitted every row — the filter silently did nothing.

**Function calls** were split on the operator, and the operand resolver separates variable from property by locating a dot. In `toUpper(a.name)` it finds the dot *inside the call* and looks up a variable named `toUpper(a`, which never resolves — so the comparison was false for every row and the filter excluded everything.

## What changed

- Label tests are parsed and evaluated against the bound node, including the `n:A:B` conjunction form.
- Label tests inside boolean combinations (`n:A OR n:B`) are substituted as boolean literals before the predicate goes to the general expression evaluator, which understands the operators but not the `var:Label` form. Substitution skips quoted regions so a colon inside a string literal is left alone.
- Predicates containing a function call are routed to the general evaluator ahead of the string-surgery branches that cannot parse them.

**Both evaluators needed it.** `evaluateWithWhereCondition` (`match_rows.go`) serves the simple-MATCH path and `evaluateWhereOnComputedRow` (`match_with_rel.go`) serves path traversal. Fixing only the first left this still unfiltered, which is the shape a scoped label whitelist actually uses:

```cypher
MATCH (s:Repository {id:'r1'})-[*1..3]->(impacted)
WITH impacted WHERE impacted:Workload OR impacted:CloudResource
RETURN impacted.id
```

## Deliberately not changed

The pass-through for wholly unrecognised predicates is pinned as intentional by `TestEvaluateWithWhereCondition_DirectBranches` ("Unknown pattern defaults to pass-through"), so it is left alone and documented in place instead.

It is worth flagging that it makes an unsupported filter behave as *no filter* rather than as an error, which is how both defects here stayed invisible. Whether that default should change seems like a maintainer decision rather than something a bug-fix PR should settle, so I have not touched it — happy to follow up either way.

## Verification

`pkg/cypher`, `pkg/storage` and `pkg/bolt` all pass. Five regression tests added, each stating the Neo4j-measured expectation in its comment so a future divergence is legible without re-deriving the reference.

Two related defects found while investigating are filed separately with reproductions and are **not** addressed here: #297 (`collect()` after two `MATCH` clauses returns no rows) and #298 (subscripting a collected list loses the node). I did not attempt those — I do not know those execution paths well enough to scope a fix confidently.

## AGENTS.md compliance

| Requirement | Result |
| --- | --- |
| Test-driven bug fix (failing test → verify fails → fix → verify passes → regression tests) | Followed; the five regression tests were written and confirmed failing before any change |
| Coverage ≥95% for core logic (`cypher`) | **100%** on all nine new/changed helpers |
| Edge cases (nil, empty, boundary) | Malformed label tests, non-node bindings, nil node/edge, missing bindings, empty and nil value maps |
| Error paths | Non-matching predicates, absent properties, unresolvable variables |
| Concurrency | 64 goroutines over the helper, 32 concurrent end-to-end queries |
| `go test -race` | Clean, 0 data races |
| Table-driven subtests | Yes, in the structure AGENTS.md documents |
| File size <2500 lines | `match_rows.go` 1132, `match_with_rel.go` 1302, tests 92 and 322 |
| `gofmt` | Clean |
| `go vet` | No findings in these files (the package's pre-existing `sync.Once` copy warnings in `cypher_helpers_extra_test.go` are on `main` and untouched here) |

Per-function coverage:

```
evaluateWithWhereCondition      100.0%
withWhereNeedsFullEvaluator     100.0%
parseWithWhereLabelTest         100.0%
withWhereNodeHasAllLabels       100.0%
withWhereIsLabelTest            100.0%
substituteWithWhereLabelTests   100.0%
isWithWhereIdentStart           100.0%
isWithWhereIdentPart            100.0%
withWhereValueContext           100.0%
```

One unreachable guard was removed rather than tested: `parseWithWhereLabelTest` checked `len(parts) < 2` after `withWhereIsLabelTest` had already established a colon with non-empty segments either side, so it could not be reached. A comment records why indexing is safe.

### CODE REVIEW

File: pkg/cypher/match_with_rel.go [hunk 1]
```diff
@@ -726,6 +726,14 @@ func (e *StorageExecutor) evaluateWhereOnComputedRow(ctx context.Context, whereC
...
 
+	// Handle a bare label test such as `n:Workload` or `n:A:B`. It carries no
+	// comparison operator, so without this branch it falls through to the
+	// pass-through at the end of this function and admits every row -- the
+	// filter is silently not applied.
+	if variable, labels, ok := parseWithWhereLabelTest(whereClause); ok {
+		return withWhereNodeHasAllLabels(values[variable], labels)
+	}
+
 	// Handle comparison operators
```
[the remark this answers is not in the document]
Comment (linuxdynasty) [reply pkg/cypher/match_with_rel.go:748]: This one does not reproduce, so I have not changed the code.

The reading rests on `evaluateExpressionFromValues` returning when its property-access branch fails to resolve the bogus varName `toUpper(x`. It does not return. At `match_with_rel.go:814-841` the `if val, ok := values[varName]; ok {` block has no `return` on the miss, so control falls through past the map-literal branch to the function-call handler at `:849`, which evaluates the call correctly.

Measured rather than argued. I reverted the *entire* path-traversal fix under `go test -overlay` and ran a table of predicates against a variable-length expansion:

```
PASS  function_call_selects_one           toUpper(x.name) = CHECKOUT  -> 1
PASS  function_call_that_matches_nothing  toUpper(x.name) = NOPE      -> 0
FAIL  label_test_selects_one              (label cases fail, as expected)
```

The second line settles it: a silent pass-through admits every row, so a predicate that correctly returns 0 rows cannot be passing through. Function calls were never broken on this path; only label tests were, which is what the fix addressed.

You did send me somewhere useful, though. Probing this turned up a real defect of the same class one line away: `WITH x WHERE NOT x:Workload` returned every row, because the commit added the bare `n:Label` branch but not its negated form, which equally has no comparison operator to split on. Fixed in a follow-up commit with a failing-then-green test, scoped to a bare label test only — negating every `NOT ...` predicate would negate the pass-through too, turning admits-every-row into drops-every-row, which is the more damaging direction.

File: pkg/cypher/match_rows.go [hunk 2]
```diff
@@ -241,7 +269,195 @@ func (e *StorageExecutor) evaluateWithWhereCondition(ctx context.Context, whereC
...
+		case c == '\'' && !inDouble:
+			inSingle = !inSingle
+		case c == '"' && !inSingle:
+			inDouble = !inDouble
+		}
+		if inSingle || inDouble {
+			out.WriteByte(c)
+			i++
+			continue
+		}
```
[the remark this answers is not in the document]
Comment (linuxdynasty) [reply pkg/cypher/match_rows.go:391]: Correct, and it reproduces exactly as described. Fixed at `pkg/cypher/match_rows.go:383-390`.

With the predicate `n.name = 'a\'b:c' OR n:Workload`, before the fix:

```
expected: n.name = 'a\'b:c' OR true
actual  : n.name = 'a\'false' OR n:Workload
```

The meaning changed twice over, which is worse than the report implies: `b:c` inside the literal was rewritten to `false`, changing the value being compared; and the genuine `n:Workload` after the `OR` was left unrewritten, so the general evaluator never saw a form it understands. The double-quoted variant fails identically.

The scanner now consumes a backslash together with a following backslash, single quote, or double quote — exactly the escape set `decodeCypherQuotedString` supports at `string_literal.go:51-58`.

I checked the premise that the two scanners must agree rather than assuming it: `substituteWithWhereLabelTests` output flows to `evaluateExpressionBoolWithContext` and on to `decodeCypherQuotedString`, so it is one string read by two scanners. The agreement is a real requirement, not a style preference.

One thing worth recording: doubled quotes already worked, but by accident — the toggle closes on the first and reopens on the second, so the enclosed text stays inside a quoted region either way. Pinned with its own case so it cannot be lost silently.

File: pkg/cypher/with_where_filter_bugs_test.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+	ctx := context.Background()
+	setupWithWhereFixture(t, store)
+
+	result, err := exec.Execute(ctx,
+		`MATCH (n) WITH n WHERE n:Workload OR n:CloudAction RETURN count(*) AS c`, nil)
+	require.NoError(t, err)
+	require.Len(t, result.Rows, 1)
+	require.EqualValues(t, 2, result.Rows[0][0],
+		"a disjunction of label tests must filter; Neo4j 5 returns 2")
+}
```
[the remark this answers is not in the document]
Comment (linuxdynasty) [reply pkg/cypher/with_where_filter_bugs_test.go:121]: Added at `pkg/cypher/with_where_filter_bugs_test.go:92-146` — table-driven over a variable-length `-[*1..2]->` expansion, eight cases: unfiltered control, label test, label-no-match, negated label, label disjunction, function-call match, function-call no-match, and a plain comparison.

One design point that matters more than the coverage itself. The two tests this replaces asserted only "1 row matches", and that shape passes under a total pass-through too — so they could not have caught the bug they were written for. Every predicate in the new table is paired with one selecting a different count, which is what makes it discriminating rather than merely present.

Proven non-vacuous: with `match_with_rel.go` reverted to the pre-fix revision under `go test -overlay`, the three label cases FAIL and the rest pass, exit 1. So the table pins the fix instead of passing regardless of it.

Writing it also surfaced a real defect this PR had left behind: `WITH x WHERE NOT x:Workload` returned every row on the traversal path, because the original commit added the bare `n:Label` branch but not its negated form, which equally has no comparison operator to split on. That is now fixed in `match_with_rel.go:737-751` with its own failing-then-green case, and it is in the table above.

File: pkg/cypher/with_where_filter_units_test.go [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+		{"matching label becomes true", "n:Workload", "true"},
+		{"non-matching label becomes false", "n:Absent", "false"},
+		{"disjunction of both", "n:Workload OR n:Absent", "true OR false"},
+		{"conjunction of labels on one variable", "n:Workload:Absent", "false"},
+		{"unknown variable is false", "missing:Workload", "false"},
+		{"non-node binding is false", "scalar:Workload", "false"},
+		{"predicate with no label test is unchanged", "n.name = 'x'", "n.name = 'x'"},
+		{"colon inside a single-quoted string is left alone", "n.name = 'a:b'", "n.name = 'a:b'"},
+		{"colon inside a double-quoted string is left alone", "n.name = \"a:b\"", "n.name = \"a:b\""},
+		{"empty predicate", "", ""},
```
[the remark this answers is not in the document]
Comment (linuxdynasty) [reply pkg/cypher/with_where_filter_units_test.go:146]: Added at `pkg/cypher/with_where_filter_units_test.go:151-160` — five cases: escaped single quote, escaped double quote, an escaped backslash correctly ending the string, doubled quotes, and a trailing backslash at end of input.

Proven non-vacuous: with `match_rows.go` reverted under `go test -overlay`, the two escape cases FAIL, exit 1. The other three are boundary pins and correctly pass either way, which is what I would want from them.

Worth recording an error I made writing these, since it is the kind that produces a green test proving nothing: my first cut used two backslashes where Cypher needs one, so the fixtures failed for the wrong reason. Caught before the fix landed — the red evidence on the linked thread comes from the corrected fixtures, not the broken ones.
